docs/evm_compatibility/overview.md

```diff
@@ -1,63 +1,155 @@
+<!-- based on item in Slack https://neonlabsworkspace.slack.com/archives/C03CQ8A6WTT/p1683107335962669 >> not sure that this is functional end user support as it stands -->
+
+When a contract requires storage slots within Solana, it may create Solana accounts and access these random addresses. However, this must be done with an awareness of the upper limit on the number of accounts, as per the next section. 
+
+
+### Upper limit on number of accounts
+Neon EVM uses [Solana Transaction V0](https://docs.solana.com/developing/versioned-transactions): limiting the maximum number of accounts used in a single transaction to 64. Solana requires that all accounts used in a transaction be specified in order; to ensure parallel execution of transactions.
+
+<!-- go deeper on HOW to modify the contract to constrain account numbers Anton will pass in slack
+  -->
+
+By constructing the contract logic differently, fixed-sized values and arrays can fit into a significantly smaller number of accounts.
+
+```Solidity
+ /*1 account: */
+uint256[] list;
+...
+
+for (uint256 i = 0; i < 32; i++) {
+   list[i] = i;
+}
+```
+
+```Solidity
+/* 32 accounts: */
+mapping(uint256 => uint256) map;
+...
+
+for (uint256 i = 0; i < 32; i++) {
+   map[i] = i;
+}
+```
+
+<!-- todo looks like these code snippets will benefit from some better commentary -->
 
 ### Heap size
-Ethereum-like transactions are executed by Neon EVM inside [Solana's Berkeley Packet Filter (BPF)](https://docs.solana.com/developing/on-chain-programs/overview#berkeley-packet-filter-bpf). The BPF has heap memory limit of 256 KB. Consequently, the size of the heap allocated to a contract call, is limited to the same 256 KB.
+Ethereum-like transactions are executed by Neon EVM inside [Solana's Berkeley Packet Filter (BPF)](https://docs.solana.com/developing/on-chain-programs/overview#berkeley-packet-filter-bpf). The BPF has heap memory limit of 256 KB, i.e. the size of the heap allocated to a contract call, is limited to 256 KB.
+
+Consider the following techniques if you need to troubleshoot a heap overflow error:
 
-To avoid the occurrence of a heap overflow error, it is recommended that you:
-* Reduce the size of the call stack
-* Reduce the number of variables used
+- Transactions: reduce the calldata size
+- Deploying/calling contracts, reduce the:
+	- Call depth
+	- Contract binary size
+	- Size of the function/constructor arguments
+- Event emission
+	- Use indexed parameters
+	- Avoid strings, arrays, and mappings
+- Returning values from a function
+	- Don't return data you don't use!
+	- Avoid strings, arrays, and mappings
+- Local variables: avoid strings, arrays, and mappings
 
-### Limitation on `block.timestamp` / `block.number` Usage
+<!-- todo How do users know when the issue is a heap overflow error? Can we show logs ?screenshot and name of service? to demonstrate when the issue is heap size? == detection method ?? Oleg should be able to provide -->
```

Comment:
@anton-lisanin Oleg could not provide anything -- when we had the meet, we talked about helping end users figure out when the issue could be attributed to heap overflow

Reply:
Example transaction with heap overflow - https://neonscan.org/tx/0x6bdc50921eaa8981e2f534e09139840fd176bf092bd835393ee4bab998409c0a

Neonscan has a list of Solana transactions that was executed. And if open last Solana transaction in solscan, you could see error message in the logs. https://solscan.io/tx/3fuTXn8SiAwTrpG96VmhvuHQe9tssFv7y2yHFynVpM4d51uBZerehxZebPTFvKLs8b2wmnXmhxu5yhAwWdnUJLVC?cluster=devnet

Reply:
@anton-lisanin So, the error they will see is Bump Allocator out of Memory?

![image](https://github.com/neonlabsorg/neon-evm.docs/assets/20266893/e1365540-8375-41c6-aa60-71f57bb2071d)

Reply:
Structure required to support end user:

- Describe Solana accounts structure
- Describe how nEVM uses accounts/ tx counter / data storage etc
{special architecture specific to nEVM -- no hash table as per Eth-EVM}
- How contract execution data is managed and held
- How Solidity contract compiler deals with batches of data in Solana accounts